### PR TITLE
Primary Visibility and Area Shadow (plus some additional raytracing features)

### PR DESCRIPTION
**Pull request by anisunity**
Adding a Raytracing render queue for both transparent and opaque objects
Opaque Objects that are in the raytracing queue are rendered in the prepass and not in the deferred/forward passes
Adding the primary visibility mode in the raytracing environement
Correcting a label in the rt_env_inspector
Adding a Debug Mode for raytracing area shadow
Adding an equivalent of the ssr lit brdf evaluation
Adding a forward raytracing pass for the lit and the unlit
Supporting Ior for the lit.shader
Implementing a HDRaytracingRenderer that handles primary visibility code
Adding resources for the raytracing primary ray
Not Pushing a light to the Light cluster if null (has been deleted)
Enabling Automatic raytracing scene invalidation on hierarchy change in the editor
Using the global command buffer parameter instead of the raytracing one when required for closest hit
Adding a texture for the analytic value (required for shadow filtering)
Adding default anyhit shaders to ambient occlusion and area shadow raytracing shaders
Adding the remainingDepth to the rayintersection structure
Adding an input reflection and transmission value for the Raytracing Lightloop and adding them to the evaluation
Adding the light dimension to the spherical rectangle
Correcting the bias usage for Raytraced AO and Reflection

### CODE REVIEW

**Comment by anisunity** [comment]
@nerites last commit!

**Comment by nerites** [comment]
Weirdly, the area shadows are still failing to render on my machine. However, re-adding the change I made doesn't seem to fix it, so I think I'll just have to figure out exactly where it's going wrong separately. Aside from the minor control flow changes (https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/2773) I'm fine with this. (:

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.hlsl [hunk 1]
```diff
@@ -1685,6 +1685,28 @@ IndirectLighting EvaluateBSDF_ScreenspaceRefraction(LightLoopContext lightLoopCo
...
+                                                    float3 transmittedColor)
+{
+    IndirectLighting lighting;
+    ZERO_INITIALIZE(IndirectLighting, lighting);
+
+#if HAS_REFRACTION
+    float3 preLD = transmittedColor;
+
+    // Inverse pre-exposure
+    float exposure = GetCurrentExposureMultiplier();
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.hlsl]: GetInverseCurrentExposureMultiplier()

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.hlsl]: GuardBand this code

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.hlsl [hunk 2]
```diff
@@ -1685,6 +1685,28 @@ IndirectLighting EvaluateBSDF_ScreenspaceRefraction(LightLoopContext lightLoopCo
     return lighting;
 }
 
+IndirectLighting EvaluateBSDF_RaytracedRefraction(LightLoopContext lightLoopContext,
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.hlsl]: GuardBand EvaluateBSDF_RaytracedRefraction

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderQueue.cs [hunk 3]
```diff
@@ -16,6 +16,9 @@ public enum Priority
         {
             Background = UnityEngine.Rendering.RenderQueue.Background,
 
+            RaytracingTransparent = UnityEngine.Rendering.RenderQueue.Geometry - 2,
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderQueue.cs]: let's discuss + comment . See with me

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderQueue.cs [hunk 4]
```diff
@@ -118,6 +129,8 @@ public static int ChangeType(RenderQueueType targetType, int offset = 0, bool al
                     return (int)Priority.LowTransparent + offset;
                 case RenderQueueType.AfterPostprocessTransparent:
                     return (int)Priority.AfterPostprocessTransparent + offset;
+                case RenderQueueType.Raytracing:
+                    return alphaTest ? (int)Priority.RaytracingTransparent : (int)Priority.RaytracingOpaque;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderQueue.cs]: please see with me

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRaytracingManager.cs [hunk 5]
```diff
@@ -394,9 +433,10 @@ public void BuildSubSceneStructure(ref HDRayTracingSubScene subScene)
                             for(int meshIdx = 0; meshIdx < numSubMeshes; ++meshIdx)
                             {
                                 Material currentMaterial = currentRenderer.sharedMaterials[meshIdx];
+                                bool materialIsTransparent = currentMaterial.IsKeywordEnabled("_SURFACE_TYPE_TRANSPARENT");
                                 if(currentMaterial != null)
                                 {
-                                    subMeshFlagArray[meshIdx] = !currentMaterial.IsKeywordEnabled("_SURFACE_TYPE_TRANSPARENT");
+                                    subMeshFlagArray[meshIdx] = true; // !currentMaterial.IsKeywordEnabled("_SURFACE_TYPE_TRANSPARENT");
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRaytracingManager.cs]: _SURFACE_TYPE_TRANSPARENT is not a keyword in shader graph, you can't rely on this...

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitRaytracing.hlsl [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+                                                    float3 transmittedColor)
+{
+    IndirectLighting lighting;
+    ZERO_INITIALIZE(IndirectLighting, lighting);
+
+#if HAS_REFRACTION
+    float3 preLD = transmittedColor;
+
+    // Inverse pre-exposure
+    preLD *= GetInverseCurrentExposureMultiplier();
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitRaytracing.hlsl]: we will need to discuss about this. for raytracing we shouldn't require this before the end.

**Pull request merged**